File: bratek20-architecture/src/main/java/pl/bratek20/architecture/properties/impl/PropertiesImpl.java
package pl.bratek20.architecture.properties.impl;

import pl.bratek20.architecture.properties.api.PropertiesApi;
import pl.bratek20.architecture.properties.api.PropertiesSource;

import java.util.List;

public class PropertiesImpl implements PropertiesApi {
    private final List<PropertiesSource> sources;

    public PropertiesImpl(List<PropertiesSource> sources) {
        this.sources = sources;
    }

    @Override
    public <T> T get(String propertyName, Class<T> propertyType) {
        return sources.get(0).get(propertyName, propertyType);
    }

    @Override
    public <T> List<T> getList(String propertyName, Class<T> propertyType) {
        return sources.get(0).getList(propertyName, propertyType);
    }
}

File: bratek20-architecture/src/main/java/pl/bratek20/architecture/properties/impl/PropertiesImpl.kt
package pl.bratek20.architecture.properties.impl

import pl.bratek20.architecture.properties.api.*

class PropertiesImpl(
    private val sources: List<PropertiesSource>
) : Properties {
    override fun <T> get(source: PropertiesSourceName, key: PropertyKey, type: Class<T>): T {
        return sources.find { it.getName() == source }?.get(key, type) ?: throw PropertiesSourceNotFoundException(source.value)
    }

    override fun <T> getList(source: PropertiesSourceName, key: PropertyKey, type: Class<T>): List<T> {
        return sources.find { it.getName() == source }?.getList(key, type) ?: throw PropertiesSourceNotFoundException(source.value)
    }
}
